File: sol/solution/0001-0100/0032. Longest Valid Parentheses/Solution.php
class Solution {
    /**
     * @param String $s
     * @return Integer
     */
    function longestValidParentheses($s) {
        $n = strlen($s); $ans = 0; $left = 0; $right = 0;
        for ($i = 0; $i < $n; $i++) {
            if ($s[$i] === '(') $left++; else $right++;
            if ($left === $right) $ans = max($ans, 2 * $right);
            else if ($right > $left) { $left = $right = 0; }
        }
        $left = $right = 0;
        for ($i = $n - 1; $i >= 0; $i--) {
            if ($s[$i] === ')') $right++; else $left++;
            if ($left === $right) $ans = max($ans, 2 * $left);
            else if ($left > $right) { $left = $right = 0; }
        }
        return $ans;
    }
}

File: sol/solution/0001-0100/0032. Longest Valid Parentheses/Solution.go
func longestValidParentheses(s string) int {
    ans, left, right := 0, 0, 0
    for i := 0; i < len(s); i++ {
        if s[i] == '(' { left++ } else { right++ }
        if left == right {
            if 2*right > ans { ans = 2 * right }
        } else if right > left {
            left, right = 0, 0
        }
    }
    left, right = 0, 0
    for i := len(s) - 1; i >= 0; i-- {
        if s[i] == ')' { right++ } else { left++ }
        if left == right {
            if 2*left > ans { ans = 2 * left }
        } else if left > right {
            left, right = 0, 0
        }
    }
    return ans
}